shell/platform/android/io/flutter/embedding/engine/systemchannels/LifecycleChannel.java

```diff
@@ -5,39 +5,95 @@
+ * windows, and combines that to decide if the application is inactive or resumed.
+ */
 public class LifecycleChannel {
   private static final String TAG = "LifecycleChannel";
+  private static final String CHANNEL_NAME = "flutter/lifecycle";
+
+  // These should stay in sync with the AppLifecycleState enum in the framework.
+  private static final String RESUMED = "AppLifecycleState.resumed";
+  private static final String INACTIVE = "AppLifecycleState.inactive";
+  private static final String PAUSED = "AppLifecycleState.paused";
+  private static final String DETACHED = "AppLifecycleState.detached";
+
+  private String lastAndroidState = "";
+  private String lastFlutterState = "";
+  private boolean lastFocus = false;
 
   @NonNull public final BasicMessageChannel<String> channel;
 
   public LifecycleChannel(@NonNull DartExecutor dartExecutor) {
-    this.channel =
-        new BasicMessageChannel<>(dartExecutor, "flutter/lifecycle", StringCodec.INSTANCE);
+    this.channel = new BasicMessageChannel<>(dartExecutor, CHANNEL_NAME, StringCodec.INSTANCE);
   }
 
-  public void appIsInactive() {
-    Log.v(TAG, "Sending AppLifecycleState.inactive message.");
-    channel.send("AppLifecycleState.inactive");
+  // Here's the state table this implements:
+  //
+  // | Android State | Window focused | Flutter state |
+  // |---------------|----------------|---------------|
+  // | Resumed       |     true       |    resumed    |
+  // | Resumed       |     false      |    inactive   |
+  // | Paused        |     true       |    inactive   |
+  // | Paused        |     false      |    inactive   |
+  // | Stopped       |     true       |    paused     |
+  // | Stopped       |     false      |    paused     |
+  // | Detached      |     true       |    detached   |
+  // | Detached      |     false      |    detached   |
+  private void sendState(String state, boolean hasFocus) {
+    if (lastAndroidState == state && hasFocus == lastFocus) {
+      // No inputs changed, so Flutter state could not have changed.
+      return;
+    }
+    String newState;
+    if (state == RESUMED) {
+      // Focus is only taken into account when the Android state is "Resumed".
+      // In all other states, focus is ignored, because we can't know what order
+      // Android lifecycle notifications and window focus notifications events
+      // will arrive in, and those states don't send input events anyhow.
+      newState = hasFocus ? RESUMED : INACTIVE;
+    } else {
+      newState = state;
+    }
+    // Keep the last reported values for future updates.
+    lastAndroidState = state;
+    lastFocus = hasFocus;
+    if (newState == lastFlutterState) {
+      // No change in the resulting Flutter state, so don't report anything.
+      return;
+    }
+    channel.send(newState);
+    lastFlutterState = newState;
+  }
+
+  public void aWindowIsFocused() {
+    sendState(lastAndroidState, true);
+  }
+
+  public void noWindowsAreFocused() {
```

Comment:
In particular, when switching from one activity to another in split-screen mode, I wonder if the `onWindowFocusChanged(false)` on the old activity comes before or after the `onWindowFocusChanged(true)` on the new one (or if both orderings are possible.)

If the `false` always comes before the `true`, then these method names are accurate. But if the `true` can come before the `false`, then on that belated `false` we'll call `noWindowsAreFocused` when in fact one window is focused, and set the Flutter state to "inactive".

Reply:
Two activities using the same flutter engine is likely to not work very well or at all, because there's no way to differentiate the Flutter UI between the two activities. And if the activities have different flutter engines, they should also have different LifecycleChannels. It's likely not an edge case you'd need to worry about.

Reply:
The intent of the "inactive" state in the framework is how I named the functions. I suspect that once we do support multiple windows in the same app (it is coming, but probably coming to mobile platforms last), there will have to be some work done to give the right information to these functions where we check on focus change and see if greater than zero windows have focus before calling one function or the other. Then the difference in order will only mean that either we notify of loss and then immediate regain of focus, or we don't notify at all, since there were never zero windows focused.

Reply:
> I suspect that once we do support multiple windows in the same app (it is coming, but probably coming to mobile platforms last), there will have to be some work done to give the right information to these functions where we check on focus change and see if greater than zero windows have focus before calling one function or the other.

Cool, that plan sounds good to me.

I guess my suggestion then would just be to either name the functions in the singular to reflect the current assumptions, or add a comment here explaining how the names reflect a future that's planned but hasn't yet arrived (and therefore doesn't yet call for logic to handle).

Reply:
I'll add a comment: I don't want to have to change the API later, and they're still accurate names even if the max number of windows is one.